File: ureka-framework-main/ureka_framework/logic/device_controller.py
# Deployment Environment
from ureka_framework.environment import Environment

# Data Model (RAM)
from typing import Optional
from ureka_framework.model.shared_data import SharedData
from ureka_framework.model.data_model.this_device import ThisDevice
from ureka_framework.model.data_model.other_device import OtherDevice
from ureka_framework.model.data_model.current_session import CurrentSession
from ureka_framework.model.data_model.this_person import ThisPerson

# Data Model (Message)
from ureka_framework.model.message_model.u_ticket import UTicket
from ureka_framework.model.message_model.r_ticket import RTicket

# Resource (Storage)
from ureka_framework.resource.storage.simple_storage import SimpleStorage

# Resource (Simulated Comm)
from ureka_framework.resource.communication.simulated_comm.simulated_comm_channel import (
    SimulatedCommChannel,
)

# Resource (Bluetooth Comm)
try:
    HAS_PYBLUEZ = True
    from ureka_framework.resource.communication.bluetooth.bluetooth_service import (
        AcceptSocket,
        ConnectingWorker,
        ConnectionSocket,
    )
except ImportError:
    HAS_PYBLUEZ = False
    # raise RuntimeError(
    #     "PyBlueZ not found - only support SIMULATED comm but not BLUETOOTH comm"
    # )

# Resource (Logger)
from ureka_framework.resource.logger.simple_logger import simple_log

# Threading
import time
import threading
from queue import Queue

# Measure Helper
from ureka_framework.logic.stage_worker.measure_helper import MeasureHelper

# Stage Worker
from ureka_framework.logic.stage_worker.msg_receiver import MsgReceiver
from ureka_framework.logic.stage_worker.received_msg_storer import ReceivedMsgStorer
from ureka_framework.logic.stage_worker.msg_verifier import MsgVerifier
from ureka_framework.logic.stage_worker.executor import Executor
from ureka_framework.logic.stage_worker.msg_generator import MsgGenerator
from ureka_framework.logic.stage_worker.generated_msg_storer import GeneratedMsgStorer
from ureka_framework.logic.stage_worker.msg_sender import MsgSender

# Pipeline Flow
from ureka_framework.logic.pipeline_flow.flow_issue_u_ticket import FlowIssueUTicket
from ureka_framework.logic.pipeline_flow.flow_apply_u_ticket import FlowApplyUTicket
from ureka_framework.logic.pipeline_flow.flow_open_session import FlowOpenSession
from ureka_framework.logic.pipeline_flow.flow_issue_u_token import FlowIssueUToken


class DeviceController:
    def __init__(self, device_type: str = None, device_name: str = None) -> None:
        # Data Model (RAM)
        self.shared_data = SharedData(
            this_device=ThisDevice(),
            current_session=CurrentSession(),
            this_person=ThisPerson(),
            device_table={},
            state=None,
        )

        # Resource (Storage)
        self.simple_storage = SimpleStorage(device_name=device_name)

        # Resource (Simulated Comm)
        self.shared_data.simulated_comm_channel = SimulatedCommChannel(
            end=None, receiver_queue=Queue(), sender_queue=None
        )
        # Resource (Bluetooth Comm)
        if HAS_PYBLUEZ == True:
            self.shared_data.accept_socket: Optional[AcceptSocket] = None
            self.shared_data.connecting_worker: Optional[ConnectingWorker] = None
            self.shared_data.connection_socket: Optional[ConnectionSocket] = None

        # Measurer
        self.shared_data.measure_rec = dict()
        self.measure_helper = MeasureHelper(shared_data=self.shared_data)

        # Stage Worker
        self.received_msg_storer = ReceivedMsgStorer(
            shared_data=self.shared_data,
            measure_helper=self.measure_helper,
            simple_storage=self.simple_storage,
        )
        self.msg_verifier = MsgVerifier(
            shared_data=self.shared_data,
            measure_helper=self.measure_helper,
        )
        self.executor = Executor(
            shared_data=self.shared_data,
            measure_helper=self.measure_helper,
            simple_storage=self.simple_storage,
            msg_verifier=self.msg_verifier,
        )
        self.msg_generator = MsgGenerator(
            shared_data=self.shared_data,
            measure_helper=self.measure_helper,
        )
        self.generated_msg_storer = GeneratedMsgStorer(
            shared_data=self.shared_data,
            measure_helper=self.measure_helper,
            simple_storage=self.simple_storage,
        )
        self.msg_sender = MsgSender(
            shared_data=self.shared_data,
            measure_helper=self.measure_helper,
        )

        # Flow
        self.flow_issuer_issue_u_ticket = FlowIssueUTicket(
            share_data=self.shared_data,
            measure_helper=self.measure_helper,
            received_msg_storer=self.received_msg_storer,
            msg_verifier=self.msg_verifier,
            executor=self.executor,
            msg_generator=self.msg_generator,
            generated_msg_storer=self.generated_msg_storer,
            msg_sender=self.msg_sender,
        )
        self.flow_open_session = FlowOpenSession(
            share_data=self.shared_data,
            measure_helper=self.measure_helper,
            received_msg_storer=self.received_msg_storer,
            msg_verifier=self.msg_verifier,
            executor=self.executor,
            msg_generator=self.msg_generator,
            generated_msg_storer=self.generated_msg_storer,
            msg_sender=self.msg_sender,
        )
        self.flow_apply_u_ticket = FlowApplyUTicket(
            share_data=self.shared_data,
            measure_helper=self.measure_helper,
            received_msg_storer=self.received_msg_storer,
            msg_verifier=self.msg_verifier,
            executor=self.executor,
            msg_generator=self.msg_generator,
            generated_msg_storer=self.generated_msg_storer,
            msg_sender=self.msg_sender,
            flow_open_session=self.flow_open_session,
        )
        self.flow_issue_u_token = FlowIssueUToken(
            share_data=self.shared_data,
            measure_helper=self.measure_helper,
            received_msg_storer=self.received_msg_storer,
            msg_verifier=self.msg_verifier,
            executor=self.executor,
            msg_generator=self.msg_generator,
            generated_msg_storer=self.generated_msg_storer,
            msg_sender=self.msg_sender,
            flow_apply_u_ticket=self.flow_apply_u_ticket,
        )

        # Stage Worker
        self.msg_receiver = MsgReceiver(
            shared_data=self.shared_data,
            measure_helper=self.measure_helper,
            msg_verifier=self.msg_verifier,
            executor=self.executor,
            msg_sender=self.msg_sender,
            flow_issuer_issue_u_ticket=self.flow_issuer_issue_u_ticket,
            flow_apply_u_ticket=self.flow_apply_u_ticket,
            flow_open_session=self.flow_open_session,
            flow_issue_u_token=self.flow_issue_u_token,
        )

        # Always load Storage after Reboot
        (
            self.shared_data.this_device,
            self.shared_data.device_table,
            self.shared_data.this_person,
            self.shared_data.current_session,
        ) = self.simple_storage.load_storage()

        # Set Device Type (must after loading storage)
        if self.shared_data.this_device.has_device_type is False:
            self.executor._execute_one_time_set_time_device_type_and_name(
                device_type, device_name
            )

        # Set Intialized State
        self.executor._intialize_state()

        simple_log("info", f"+ Here is a {self.shared_data.this_device.device_name}...")

    ######################################################
    # Device Activity Cycle
    ######################################################
    def reboot_device(self) -> None:
        simple_log("info", f"+ Reboot {self.shared_data.this_device.device_name}...")
        self.__init__(
            device_type=self.shared_data.this_device.device_type,
            device_name=self.shared_data.this_device.device_name,
        )

File: UrekaApp/app/src/main/java/ureka/framework/logic/DeviceController.java
package ureka.framework.logic;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.example.urekaapp.ble.BLEViewModel;
import com.example.urekaapp.ble.BLEManager;
import com.example.urekaapp.communication.NearbyManager;
import com.example.urekaapp.communication.NearbyViewModel;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonSyntaxException;

import java.util.HashMap;
import java.util.Objects;

import ureka.framework.Environment;
import ureka.framework.logic.pipeline_flow.FlowApplyUTicket;
import ureka.framework.logic.pipeline_flow.FlowIssueUTicket;
import ureka.framework.logic.pipeline_flow.FlowIssueUToken;
import ureka.framework.logic.pipeline_flow.FlowOpenSession;
import ureka.framework.logic.stage_worker.Executor;
import ureka.framework.logic.stage_worker.GeneratedMsgStorer;
import ureka.framework.logic.stage_worker.MeasureHelper;
import ureka.framework.logic.stage_worker.MsgGenerator;
import ureka.framework.logic.stage_worker.MsgReceiver;
import ureka.framework.logic.stage_worker.MsgSender;
import ureka.framework.logic.stage_worker.MsgVerifier;
import ureka.framework.logic.stage_worker.ReceivedMsgStorer;
import ureka.framework.model.SharedData;
import ureka.framework.model.data_model.CurrentSession;
import ureka.framework.model.data_model.ThisDevice;
import ureka.framework.model.data_model.ThisPerson;
import ureka.framework.resource.communication.simulated_comm.SimulatedCommChannel;
import ureka.framework.resource.logger.SimpleLogger;
import ureka.framework.resource.storage.SimpleStorage;

public class DeviceController {
    private SharedData sharedData;
    private SimpleStorage simpleStorage;
    private MeasureHelper measureHelper;
    private ReceivedMsgStorer receivedMsgStorer;
    private MsgVerifier msgVerifier;
    private Executor executor;
    private MsgGenerator msgGenerator;
    private GeneratedMsgStorer generatedMsgStorer;
    private MsgSender msgSender;
    private MsgReceiver msgReceiver;
    private BLEManager bleManager;
    private NearbyManager nearbyManager;
    // Flow
    private FlowIssueUTicket flowIssuerIssueUTicket;
    private FlowOpenSession flowOpenSession;
    private FlowApplyUTicket flowApplyUTicket;
    private FlowIssueUToken flowIssueUToken;

    public DeviceController(String deviceType, String deviceName) {
        this._initialize(deviceType, deviceName, Environment.applicationContext);
    }
    private void _initialize(String deviceType, String deviceName, Context context) {
        BLEViewModel bleViewModel = new ViewModelProvider((AppCompatActivity) context).get(BLEViewModel.class);
        NearbyViewModel nearbyViewModel = new ViewModelProvider((AppCompatActivity) context).get(NearbyViewModel.class);

        this.sharedData = new SharedData(new ThisDevice(), new CurrentSession(), new ThisPerson());
        this.simpleStorage = new SimpleStorage(deviceName);
        this.sharedData.setSimulatedCommChannel(new SimulatedCommChannel());
//        # Resource (Bluetooth Comm)
//        if HAS_PYBLUEZ == True:
//        self.shared_data.accept_socket: Optional[AcceptSocket] = None
//        self.shared_data.connecting_worker: Optional[ConnectingWorker] = None
//        self.shared_data.connection_socket: Optional[ConnectionSocket] = None
        this.sharedData.setMeasureRec(new HashMap<>());
        this.measureHelper = new MeasureHelper(this.sharedData);

        this.receivedMsgStorer = new ReceivedMsgStorer(this.sharedData, this.measureHelper, this.simpleStorage);
        this.msgVerifier = new MsgVerifier(this.sharedData, this.measureHelper);
        this.executor = new Executor(this.sharedData, this.measureHelper, this.simpleStorage, this.msgVerifier);
        this.msgGenerator = new MsgGenerator(this.sharedData, this.measureHelper);
        this.generatedMsgStorer = new GeneratedMsgStorer(this.sharedData, this.measureHelper, this.simpleStorage);
        this.bleManager = bleViewModel.getBLEManager(context);
        this.nearbyManager = nearbyViewModel.getNearbyManager(context, this.msgReceiver);
        this.msgSender = new MsgSender(this.sharedData, this.measureHelper, this.bleManager, this.nearbyManager);

        // Flow
        this.flowIssuerIssueUTicket = new FlowIssueUTicket(
            this.sharedData, this.measureHelper, this.receivedMsgStorer, this.msgVerifier
            , this.executor, this.msgGenerator, this.generatedMsgStorer, this.msgSender
        );
        this.flowOpenSession = new FlowOpenSession(
            this.sharedData, this.measureHelper, this.receivedMsgStorer, this.msgVerifier,
            this.executor, this.msgGenerator, this.generatedMsgStorer, this.msgSender
        );
        this.flowApplyUTicket = new FlowApplyUTicket(
            this.sharedData, this.measureHelper, this.receivedMsgStorer, this.msgVerifier,
            this.executor, this.msgGenerator, this.generatedMsgStorer, this.msgSender, this.flowOpenSession
        );
        this.flowIssueUToken = new FlowIssueUToken(
            this.sharedData, this.measureHelper, this.receivedMsgStorer, this.msgVerifier,
            this.executor, this.msgGenerator, this.generatedMsgStorer, this.msgSender, this.flowApplyUTicket
        );

        this.msgReceiver = new MsgReceiver(
            this.sharedData, this.measureHelper, this.msgVerifier, this.executor, this.msgSender,
            this.flowIssuerIssueUTicket, this.flowApplyUTicket, this.flowOpenSession, this.flowIssueUToken
        );

//        SimpleStorage.Tuple tuple = this.simpleStorage.loadStorage();
//        this.sharedData.setThisDevice(tuple.thisDevice);
//        this.sharedData.setDeviceTable(tuple.deviceTable);
//        this.sharedData.setThisPerson(tuple.thisPerson);
//        this.sharedData.setCurrentSession(tuple.currentSession);
//        # Always load Storage after Reboot
//            (
//                self.shared_data.this_device,
//                self.shared_data.device_table,
//                self.shared_data.this_person,
//                self.shared_data.current_session,
//                ) = self.simple_storage.load_storage()

        // Set Device Type (must after loading storage)
        if (!this.sharedData.getThisDevice().getHasDeviceType()) {
            this.executor._executeOneTimeSetTimeDeviceTypeAndName(deviceType, deviceName);
        }
        // Set Initialized State
        this.executor._initializeState();

        SimpleLogger.simpleLog("info", "Here is a " + this.sharedData.getThisDevice().getDeviceName() + "...");
    }

    //////////////////////////////////////////////////////
    // Device Activity Cycle
    //////////////////////////////////////////////////////
    public void rebootDevice() {
        SimpleLogger.simpleLog("info", "Reboot " + this.sharedData.getThisDevice().getDeviceName() + "...");
        this._initialize(this.sharedData.getThisDevice().getDeviceType(), this.sharedData.getThisDevice().getDeviceName(), Environment.applicationContext);
    }

    public SharedData getSharedData() {
        return this.sharedData;
    }
    public void setSharedData(SharedData sharedData) {
        this.sharedData = sharedData;
    }

    public SimpleStorage getSimpleStorage() {
        return simpleStorage;
    }

    public void setSimpleStorage(SimpleStorage simpleStorage) {
        this.simpleStorage = simpleStorage;
    }

    public MeasureHelper getMeasureHelper() {
        return measureHelper;
    }

    public void setMeasureHelper(MeasureHelper measureHelper) {
        this.measureHelper = measureHelper;
    }

    public ReceivedMsgStorer getReceivedMsgStorer() {
        return receivedMsgStorer;
    }

    public void setReceivedMsgStorer(ReceivedMsgStorer receivedMsgStorer) {
        this.receivedMsgStorer = receivedMsgStorer;
    }

    public MsgVerifier getMsgVerifier() {
        return msgVerifier;
    }

    public void setMsgVerifier(MsgVerifier msgVerifier) {
        this.msgVerifier = msgVerifier;
    }

    public Executor getExecutor() {
        return executor;
    }

    public void setExecutor(Executor executor) {
        this.executor = executor;
    }

    public MsgGenerator getMsgGenerator() {
        return msgGenerator;
    }

    public void setMsgGenerator(MsgGenerator msgGenerator) {
        this.msgGenerator = msgGenerator;
    }

    public GeneratedMsgStorer getGeneratedMsgStorer() {
        return generatedMsgStorer;
    }

    public void setGeneratedMsgStorer(GeneratedMsgStorer generatedMsgStorer) {
        this.generatedMsgStorer = generatedMsgStorer;
    }

    public MsgSender getMsgSender() {
        return msgSender;
    }

    public void setMsgSender(MsgSender msgSender) {
        this.msgSender = msgSender;
    }

    public MsgReceiver getMsgReceiver() {
        return msgReceiver;
    }

    public void setMsgReceiver(MsgReceiver msgReceiver) {
        this.msgReceiver = msgReceiver;
    }

    public FlowIssueUTicket getFlowIssuerIssueUTicket() {
        return flowIssuerIssueUTicket;
    }

    public void setFlowIssuerIssueUTicket(FlowIssueUTicket flowIssuerIssueUTicket) {
        this.flowIssuerIssueUTicket = flowIssuerIssueUTicket;
    }

    public FlowOpenSession getFlowOpenSession() {
        return flowOpenSession;
    }

    public void setFlowOpenSession(FlowOpenSession flowOpenSession) {
        this.flowOpenSession = flowOpenSession;
    }

    public FlowApplyUTicket getFlowApplyUTicket() {
        return flowApplyUTicket;
    }

    public void setFlowApplyUTicket(FlowApplyUTicket flowApplyUTicket) {
        this.flowApplyUTicket = flowApplyUTicket;
    }

    public FlowIssueUToken getFlowIssueUToken() {
        return flowIssueUToken;
    }

    public void setFlowIssueUToken(FlowIssueUToken flowIssueUToken) {
        this.flowIssueUToken = flowIssueUToken;
    }

    public void connectToDevice(String deviceName, Runnable onConnected, Runnable onDisconnected, TextView textView) {
        new Thread(() -> {
            this.bleManager.startScan(deviceName, new BLEManager.BLECallback() {
                final StringBuilder jsonBuilder = new StringBuilder();
                @Override
                public void onConnected() {
                    SimpleLogger.simpleLog("info", "Device connected!");
                    new Handler(Looper.getMainLooper()).post(() ->
                            textView.setText("Device connected!")
                    );
                    onConnected.run();
                }

                @Override
                public void onDisconnected() {
                    SimpleLogger.simpleLog("info", "Device disconnected!");
                    new Handler(Looper.getMainLooper()).post(() ->
                            textView.setText("Device disconnected!")
                    );
                    onDisconnected.run();
                }

                @Override
                public void onDataReceived(String data) {
                    SimpleLogger.simpleLog("info", "Received data: " + data);
//                    new Handler(Looper.getMainLooper()).post(() ->
//                            textView.setText("Data received from voting machine.")
//                    );
                    jsonBuilder.append(data);

                    if (data.contains("freed")) {
                        SimpleLogger.simpleLog("info", "No hello thank you");
                        jsonBuilder.setLength(0);
                    } else if (data.contains("$")) {
                        Log.d("Bluetooth.onDataReceived", "Received complete JSON: starting process.");
                        jsonBuilder.deleteCharAt(jsonBuilder.length() - 1);
                        msgReceiver._recvXxxMessage(jsonBuilder.toString());
                        jsonBuilder.setLength(0);
                    } else {
                        Log.d("Bluetooth.onDataReceived", "Waiting for more data to complete JSON.");
                    }
                }
            });
        }).start();
    }
}
